distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/JvmErgonomics.java

```diff
@@ -53,25 +53,57 @@ private JvmErgonomics() {
      */
     static List<String> choose(final List<String> userDefinedJvmOptions) throws InterruptedException, IOException {
         final List<String> ergonomicChoices = new ArrayList<>();
-        final Map<String, Optional<String>> finalJvmOptions = finalJvmOptions(userDefinedJvmOptions);
+        final Map<String, JvmOption> finalJvmOptions = finalJvmOptions(userDefinedJvmOptions);
         final long heapSize = extractHeapSize(finalJvmOptions);
         final long maxDirectMemorySize = extractMaxDirectMemorySize(finalJvmOptions);
         if (maxDirectMemorySize == 0) {
             ergonomicChoices.add("-XX:MaxDirectMemorySize=" + heapSize / 2);
         }
+        final boolean tuneG1GCForSmallHeap = tuneG1GCForSmallHeap(finalJvmOptions, heapSize);
+        final boolean tuneG1GCForLargeHeap = tuneG1GCForLargeHeap(finalJvmOptions, heapSize);
+        if (tuneG1GCForSmallHeap) {
```

Comment:
I think it is surprising that if you manually specify any of the 3 parameters you loose all 3. For instance if they decide to raise `G1ReservePercent` to 20, the region size adjustment is lost.

I think the check in `tuneG1GCForSmallHeap` should only check heap size and maybe that G1 is in use.

The code below should then only add options where there is no existing `command line` origin option.

Alternatively (to the last part), we could swap the two lines here:

https://github.com/elastic/elasticsearch/blob/01dc110c7286f7e83ea92f90dad17376db089a1a/distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/JvmOptionsParser.java#L144

to be:

```
        finalJvmOptions.addAll(ergonomicJvmOptions);
        finalJvmOptions.addAll(substitutedJvmOptions);
```

since that would ensure that the original command line options override the ergonomics picked here.